Added initial guide on the TCP =>gRPC client migration

oskardudycz opened the pull request:
Added the initial migration guide including:
- framework and packages updates,
- explained migration strategies,
- differences in the connection management,
- security changes,
- operations (appends, reads) migration,
- built-in migrations,
- transactions.

It does not include:
- catch-up subscriptions. See https://github.com/EventStore/documentation/pull/498,
- persistent subscriptions.

They will come as separate, dedicated PRs.

_Note: Migration guide, will need to be copied/pasted to other TCP client docs versions. I'll do it once the changes are approved to reduce the churn._

Docs preview: https://deploy-preview-486--developers-eventstore.netlify.app/clients/dotnet/21.2/migration-to-gRPC.html

Fixes: #444

oskardudycz committed:
Added initial description on the TCP=>gRPC migration

oskardudycz committed:
Added note about transactions

oskardudycz committed:
Added information about EventData changes for appending events

oskardudycz committed:
Added information about TCP => gRPC optimistic concurrency changes

oskardudycz committed:
Updated connection management differences between TCP and gRPC based on the feedback.

oskardudycz committed:
Added read events migration guidance for TCP => gRPC client migration

oskardudycz committed:
Added basic information about TCP=>gRPC connection string and security considerations

oskardudycz committed:
Added information about retry policy migration from TCP to gRPC

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,599 @@
+        if (page.IsEndOfStream) break;
+
+        start += pageSize;
+    } while (true);
```
Can do `while (!page.IsEndOfStream`) and avoid if/break

oskardudycz replied:
Fixed in https://github.com/EventStore/documentation/pull/486/commits/ffc35a21c4c86f91e92896e4d467dece8d636351.

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,599 @@
+        StreamPosition.Start
+    );
+
+    if(await readResult.ReadState != ReadState.Ok)
```
missing space

oskardudycz replied:
Fixed in https://github.com/EventStore/documentation/pull/486/commits/ffc35a21c4c86f91e92896e4d467dece8d636351.

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,599 @@
+
+### Serialisation
+
+The gRPC client uses `ReadOnlyMemory<byte>` instead of `byte` array to make the events processing more efficient. To support that, you need to modify your deserialisation logic slightly:
```
instead of byte array or `byte[]`

oskardudycz replied:
Fixed in https://github.com/EventStore/documentation/pull/486/commits/ffc35a21c4c86f91e92896e4d467dece8d636351.

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,599 @@
+      var dataType = TypeMapper.GetType(resolvedEvent.Event.EventType);
+-     var jsonData = Encoding.UTF8.GetString(resolvedEvent.Event.Data);
++     var jsonData = Encoding.UTF8.GetString(resolvedEvent.Event.Data.Span);
+      var data = JsonConvert.DeserializeObject(jsonData, dataType);
```
Do we need to use Newtonsoft.Json in these examples? .NET serializer can directly serialize to bytes.

oskardudycz replied:
Added a disclaimer in https://github.com/EventStore/documentation/pull/486/commits/3e9d2d2757645583da3718b5b3ec8eef2c983165.

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,599 @@
+    .LimitRetriesForOperationTo(7);
+
+var connection = 
+    EventStoreConnection.Create(
```
Can be one line

oskardudycz replied:
Fixed in https://github.com/EventStore/documentation/pull/486/commits/ffc35a21c4c86f91e92896e4d467dece8d636351.

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,599 @@
+{
+    private static readonly AsyncPolicyWrap RetryPolicy
+        = Policy
+            .Handle<Exception>()
```
Not sure it's a good idea to retry everything. Like `StreamDeleted` is better not to retry.

oskardudycz replied:
Added warning for retry policy definition in: https://github.com/EventStore/documentation/pull/486/commits/ffc35a21c4c86f91e92896e4d467dece8d636351.

oskardudycz replied:
I also replaced the example of retry to reads instead of appends in https://github.com/EventStore/documentation/pull/486/commits/1225180ca1aafe17ff5cda49c18adb89d77d1e85.

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
````diff
@@ -0,0 +1,599 @@
+and to  the `$all` stream backwards:
+
+```csharp
+await using var readResult = client.ReadAllAsync(
````
Don't we want to add the max count to all of those? It will timeout if you read without max count.

oskardudycz replied:
I'd keep it like that, as I hope that the new release will be made next week, and this case will be handled correctly.

alexeyzimarev commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
````diff
@@ -0,0 +1,599 @@
+}
+```
+
+## Differences in Connection Management
````
We don't use capital casing for titles, I keep fixing it everywhere....

oskardudycz replied:
Fixed in https://github.com/EventStore/documentation/pull/486/commits/ffc35a21c4c86f91e92896e4d467dece8d636351.

oskardudycz committed:
Added more context for the TCP=>gRPC migration
Explained why Json.NET is used instead of System.Text.Json in the gRPC=>TCP migration guide

oskardudycz committed:
Improved text and code samples formatting in TCP=>gRPC migration guide.
Added warning for retry policy definition.

oskardudycz committed:
Replaced append with read in the example retry polic in TCP => gRPC migration guide.

oskardudycz committed:
Updated wording and fixed typos in TCP=>gRPC migration guide

oskardudycz committed:
Added more clarification about the TCP <=> gRPC clients connection strings' compatibility

oskardudycz committed:
Aligned TCP connection and gRPC client naming to make the difference clearer

ylorph commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+- .NET 6 (`net6.0`),
+- .NET 5 (`net5.0`),
+- .NET Core 3.1 (`netcoreapp3.1`),
+- .NET Framework 4.8 (`net48`)
```
well .net 48.....

oskardudycz replied:
Fair point 😉 Removed in https://github.com/EventStore/documentation/pull/486/commits/bfc112498e63d9366d62fc5a7519150f5c0a9a21.

ylorph commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+## Update package references
+
+To use the gRPC client, you need to replace the TPC client package (`EventStore.Client`) with the gRPC one (`EventStore.Client.Grpc.Streams`). 
```
```suggestion
To use the gRPC client, you need to replace the TCP client package (`EventStore.Client`) with the gRPC one (`EventStore.Client.Grpc.Streams`). 
```

ylorph reviewed (commented):
( still under review )

thefringeninja committed:
Update migration-to-gRPC.md

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+We decided to use gRPC as our communication protocol, as it enabled us to standardise client communication. Thanks to that, we benefit from the built-in gRPC functionalities like streaming, back-pressure,etc., instead of building custom implementations. 
+
+As gRPC is a multi-platform and widely adopted standard, it enabled us to provide a unified approach and deliver clients for other dev environments like Go, JVM, NodeJs, and Rust.
```
```suggestion
As gRPC is multi-platform and a widely adopted standard, it has enabled us to provide a unified approach and deliver clients for other dev environments like Go, JVM, NodeJs, and Rust.
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+As gRPC is a multi-platform and widely adopted standard, it enabled us to provide a unified approach and deliver clients for other dev environments like Go, JVM, NodeJs, and Rust.
+
+The TCP client will still be getting necessary bug fixes, but new database features will be only delivered for the gRPC client (e.g., persistent subscription to `$all`).
```
```suggestion
The TCP client will still get necessary bug fixes, but new database features will be only delivered for gRPC clients (e.g., persistent subscription to `$all`).
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+## Migration strategies
+
+You may also consider step by step migration:
```
```suggestion
You may also consider a step by step migration:
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+## Differences in connection management
+
+Both the TCP and gRPC clients are managing reconnections. That's the reason why it should be registered as a single instance in the application. 
```
```suggestion
Both the TCP and gRPC clients manage reconnections. Therefore it should be registered as a single instance in the application. 
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
````diff
@@ -0,0 +1,630 @@
+}
+```
+
+As TCP client connection logic is not thread-safe (`ConnectAsync` method), other operations are thread-safe. Because of that, you have to ensure that you won't have a race condition. You also should consider handling `Closed` events to manage reconnection if it is closed. 
````
```suggestion
The TCP client connection logic is not thread-safe (`ConnectAsync` method), however the other operations are thread-safe. Because of that, you have to ensure that you won't have a race condition. You also should consider handling `Closed` events to manage reconnection if it is closed. 
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+As TCP client connection logic is not thread-safe (`ConnectAsync` method), other operations are thread-safe. Because of that, you have to ensure that you won't have a race condition. You also should consider handling `Closed` events to manage reconnection if it is closed. 
+
+TCP client supports built-in reconnections; however, you need to be careful about setting the connections options properly. Wrongly defined settings can cause a flood of reconnections, increasing the chance for failure. For instance, additional retries may worsen if the reason was a high load on the database.
```
```suggestion
The TCP client supports built-in reconnections; however, you need to be careful about setting the connections options properly. Wrongly defined settings can cause a flood of reconnections, increasing the chance for failure. For instance, additional retries may worsen if the reason was a high load on the database.
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+### Transactions
+
+The most significant breaking change in the gRPC client is that it **does not support transactions anymore**. The TCP client may perform multiple appends to EventStoreDB as one transaction. The transaction can only append events to one stream. Transactions across multiple streams are not supported. The gRPC client still supports appending more than one event to the single stream as an atomic oper ...
```
```suggestion
The most significant breaking change in the gRPC client is that it **does not support transactions anymore**. The TCP client may perform multiple appends to EventStoreDB in one transaction. The transaction can only append events to one stream. Transactions across multiple streams are not supported. The gRPC client still supports appending more than one event to the single stream as an atomic operation. 
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+### Read direction
+
+The TCP Client have dedicated methods for reading events in the specific direction: 
```
```suggestion
The TCP Client has dedicated methods for reading events in the specific direction: 
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+### Read positions
+
+Both TCP and gRPC clients allow reading the stream from a specific position (representing the location of the particular event in the stream). We decided to make positions more explicit accordingly to the stream revision changes in [appending new events](#appending-events). 
```
```suggestion
Both TCP and gRPC clients allow reading the stream from a specific position (representing the location of the particular event in the stream). We decided to make positions more explicit according to the stream revision changes in [appending new events](#appending-events). 
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+Both TCP and gRPC clients allow reading the stream from a specific position (representing the location of the particular event in the stream). We decided to make positions more explicit accordingly to the stream revision changes in [appending new events](#appending-events). 
+
+We expanded a `StreamPosition` class that centralises stream position handling. It still has a `Start` and an `End` constant representing the logical stream's position. However, they return no longer `long` values but `StreamPosition` class instance. We replaced `long` values with `ulong` as stream position is always positive.
```
```suggestion
We expanded a `StreamPosition` class that centralises stream position handling. It still has a `Start` and an `End` constant representing the logical stream's position. However, they no longer return `long` values but a `StreamPosition` class instance. We replaced `long` values with `ulong` as stream position is always positive.
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+### Read result
+
+TCP Client requires paging through the results. You must provide the maximum number of events you want to read in the single read call. In the gRPC client, this is optional. By default, it will try to read all events. To make it efficient, read methods return [IAsyncEnumerable](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8 ...
```
```suggestion
The TCP Client requires paging through the results. You must provide the maximum number of events you want to read in the single read call. In the gRPC client, this is optional. By default, it will try to read all events. To make it efficient, read methods return [IAsyncEnumerable](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8). Which means that it won't load the whole stream at once but iterate sequentially, reducing the memory pressure.
```

thefringeninja commented on docs/clients/dotnet/21.2/migration-to-gRPC.md:
```diff
@@ -0,0 +1,630 @@
+
+### Serialisation
+
+The gRPC client uses `ReadOnlyMemory<byte>` instead of byte array to make event processing more efficient. To support that, you will need to modify your deserialisation logic slightly:
```
```suggestion
The gRPC client uses `ReadOnlyMemory<byte>` instead of a byte array to make event processing more efficient. To support that, you will need to modify your deserialisation logic slightly:
```

oskardudycz committed:
Apply styling suggestions from code review

Co-authored-by: João Bragança <495495+thefringeninja@users.noreply.github.com>
Co-authored-by: ylorph <ylorph@users.noreply.github.com>

oskardudycz committed:
Removed note about .NET Framework 4.8 support (as it'll be dropped in the next release)

oskardudycz committed:
Copied migration guide to TCP v5 and v20.10 client versions

oskardudycz committed:
Added note to the main TCP client README about being legacy.
